src/librustdoc/html/render/context.rs

```diff
@@ -301,13 +301,10 @@ impl<'tcx> Context<'tcx> {
     /// may happen, for example, with externally inlined items where the source
     /// of their crate documentation isn't known.
     pub(super) fn src_href(&self, item: &clean::Item) -> Option<String> {
-        self.href_from_span(item.span(self.tcx()), true)
+        self.href_from_span(item.span(self.tcx())?, true)
     }
 
     pub(crate) fn href_from_span(&self, span: clean::Span, with_lines: bool) -> Option<String> {
-        if span.is_dummy() {
-            return None;
-        }
```

Comment:
I wonder if there's any points at which we could get dummy spans from rustc itself?

Reply:
Not sure -- maybe from a really strange macro expansion? 

My guess was that this was originally placed to deal with the auto-trait issue I am fixing here more systematically, but maybe it also addresses other sources of DUMMY_SP that I didn't consider... :thinking:

Reply:
Proc macros can generate items without spans, I think.

Reply:
There's no such thing as "without span" I think? -- proc macros have a [few choices of spans](https://doc.rust-lang.org/proc_macro/struct.Span.html#method.call_site) to associate with tokens, but I can't tell how you'd get a `DUMMY_SP` out of a proc macro except during error recovery?

Reply:
TokenStream doesn't actually require a span though, right? https://doc.rust-lang.org/proc_macro/struct.TokenStream.html#method.new
Only Ident requires a span. https://doc.rust-lang.org/proc_macro/enum.TokenTree.html

Reply:
@jyn514 sure, but `TokenTree`s have spans (and presumably need them to be created), and anyways you can't create a rustdoc item with an empty `TokenStream`, right?

Reply:
Hmm, I guess so. I'm just really wary of changes here - if you're confident dummy spans are impossible I'd like to make that an assert instead of completely removing the code l

Reply:
Yeah I can do that

Reply:
Actually, this failed in rollup because we tried to call rustdoc on an empty file in a cargo unit test.

Specifically, if the file contents are empty, then the span of the root module of the crate is gonna be a span that is exactly equal to `DUMMY_SP` (because it goes from 0 to 0). So when we try to generate an href to the module sources, we trigger an assertion failure.

I think we should just remove this assertion I added. That is, if a `DUMMY_SP` (or, in this case, a span that looks just like one) comes from rustc, we should be not doing anything special to it. I think this will generate a src link that points to the empty file, but I feel like that's the simpler and easier to understand behavior.

cc @camelid @jyn514 thoughts?

Reply:
I really wish we had a way to collect metrics for rustdoc :(

but yeah removing the assertion makes sense to me, if someone reports a bug we can fix it